src/service/resources-impl.js

```diff
@@ -407,13 +410,29 @@ export class Resources {
    * @private
    */
   buildReadyResources_() {
+    // Avoid cases where elements add more elements inside of them
+    // and cause an infinite loop of building - see #3354 for details.
+    if (this.isCurrentlyBuildingPendingResources_) {
```

Comment:
hmmm. Not sure about the optimizers but yeah if it gets translated to this then we'd have the problem you mentioned. If we're worried this could actually happen, I can go ahead and use your suggestion of using a counter instead of a boolean?